src/analytics/items/personalize_item_mapper.py

```diff
@@ -0,0 +1,217 @@
+"""
+Functional mapper for converting ResearchhubUnifiedDocument to AWS Personalize items.
+"""
+
+import logging
+from typing import Dict, Optional, Protocol, runtime_checkable
+
+from django.conf import settings
+from django.db import connection
+
+from analytics.constants.personalize_constants import (
+    AUTHOR_IDS,
+    BLUESKY_COUNT_TOTAL,
+    BOUNTY_HAS_SOLUTIONS,
+    CITATION_COUNT_TOTAL,
+    CREATION_TIMESTAMP,
+    CSV_HEADERS,
+    DELIMITER,
+    HAS_ACTIVE_BOUNTY,
+    HUB_IDS,
+    HUB_L1,
+    HUB_L2,
+    ITEM_ID,
+    ITEM_TYPE,
+    PROPOSAL_HAS_FUNDERS,
+    PROPOSAL_IS_OPEN,
+    RFP_HAS_APPLICANTS,
+    RFP_IS_OPEN,
+    TEXT,
+    TITLE,
+    TWEET_COUNT_TOTAL,
+    UPVOTE_SCORE,
+)
+from analytics.utils.personalize_item_utils import assert_no_queries, clean_text_for_csv
+from utils.time import datetime_to_epoch_seconds
+
+logger = logging.getLogger(__name__)
+
+
+@runtime_checkable
+class PrefetchedUnifiedDocument(Protocol):
+    """
+    UnifiedDocument with required prefetched relations.
+
+    Required prefetch_related:
+    - hubs
+    - grants, grants__contacts__author_profile
+    - fundraises, related_bounties
+    - paper__authorships__author
+    - posts__authors
+    """
+
+    id: int
+    document_type: str
+    score: int
+
+
+def map_to_item(
```

Comment:
Instead of free-floating  functions, this could be in a mapper class that can be injected into the service. It will also make testing easier (no needed for patching).